File: src/core/account/accounts.test.ts
import { describe, expect, it } from 'vitest'
import {
  EMPTY_ACCOUNTS,
  addAccount,
  cookieDaysLeft,
  currentAccountOf,
  cycleAccountUid,
  formatCookieExpiry,
  parseStoredAccounts,
  removeAccount,
  serializeAccounts,
  switchAccount,
  type AccountsState,
  type NgaAccount,
} from './accounts'

function account(uid: string, overrides: Partial<NgaAccount> = {}): NgaAccount {
  return { uid, cid: `cid-${uid}`, name: `用户${uid}`, loginAt: 1_754_000_000_000, ...overrides }
}

const TWO: AccountsState = {
  accounts: [account('1001'), account('1002')],
  currentUid: '1001',
}

describe('addAccount', () => {
  it('新账号追加到末尾并立即成为当前账号', () => {
    const state = addAccount(EMPTY_ACCOUNTS, account('1001'))
    expect(state.accounts.map((a) => a.uid)).toEqual(['1001'])
    expect(state.currentUid).toBe('1001')

    const next = addAccount(state, account('1002'))
    expect(next.accounts.map((a) => a.uid)).toEqual(['1001', '1002'])
    expect(next.currentUid).toBe('1002')
  })

  it('同 uid 重登就地刷新 cid 与 loginAt,位置不动', () => {
    const relogin = account('1001', { cid: 'cid-new', loginAt: 1_754_100_000_000 })
    const state = addAccount(TWO, relogin)
    expect(state.accounts.map((a) => a.uid)).toEqual(['1001', '1002'])
    expect(state.accounts[0]).toEqual(relogin)
    expect(state.currentUid).toBe('1001')
  })
})

describe('switchAccount / removeAccount', () => {
  it('切换只动 currentUid;uid 不在表里原样返回', () => {
    expect(switchAccount(TWO, '1002').currentUid).toBe('1002')
    expect(switchAccount(TWO, '9999')).toBe(TWO)
  })

  it('退出当前账号落到剩余第一个;全退光即游客态(currentUid=null)', () => {
    const one = removeAccount(TWO, '1001')
    expect(one.accounts.map((a) => a.uid)).toEqual(['1002'])
    expect(one.currentUid).toBe('1002')

    const none = removeAccount(one, '1002')
    expect(none).toEqual(EMPTY_ACCOUNTS)
  })

  it('退出非当前账号不影响当前账号', () => {
    const state = removeAccount(TWO, '1002')
    expect(state.currentUid).toBe('1001')
  })

  it('currentAccountOf 取当前账号;游客态是 null', () => {
    expect(currentAccountOf(TWO)?.uid).toBe('1001')
    expect(currentAccountOf(EMPTY_ACCOUNTS)).toBeNull()
  })
})

describe('cycleAccountUid · 抽屉左右滑动切换', () => {
  const three: AccountsState = {
    accounts: [account('1001'), account('1002'), account('1003')],
    currentUid: '1002',
  }

  it('前后循环取,到头绕回', () => {
    expect(cycleAccountUid(three, 1)).toBe('1003')
    expect(cycleAccountUid(three, -1)).toBe('1001')
    expect(cycleAccountUid({ ...three, currentUid: '1003' }, 1)).toBe('1001')
    expect(cycleAccountUid({ ...three, currentUid: '1001' }, -1)).toBe('1003')
  })

  it('不足两个账号没得切', () => {
    expect(cycleAccountUid(EMPTY_ACCOUNTS, 1)).toBeNull()
    expect(cycleAccountUid({ accounts: [account('1001')], currentUid: '1001' }, 1)).toBeNull()
  })
})

describe('cookie 过期天数(30 天惯例,本地推算)', () => {
  const DAY = 24 * 60 * 60 * 1000
  const loginAt = 1_754_000_000_000

  it('刚登录 30 天,过一天半剩 29(向上取整)', () => {
    expect(cookieDaysLeft(loginAt, loginAt)).toBe(30)
    expect(cookieDaysLeft(loginAt, loginAt + 1.5 * DAY)).toBe(29)
    expect(cookieDaysLeft(loginAt, loginAt + 29.5 * DAY)).toBe(1)
  })

  it('过了 30 天算已过期', () => {
    expect(cookieDaysLeft(loginAt, loginAt + 31 * DAY)).toBeLessThanOrEqual(0)
    expect(formatCookieExpiry(loginAt, loginAt + 31 * DAY)).toBe('已过期')
    expect(formatCookieExpiry(loginAt, loginAt)).toBe('30 天后过期')
  })
})

describe('SecureStore 序列化往返与容错', () => {
  it('serialize → parse 往返一致', () => {
    expect(parseStoredAccounts(serializeAccounts(TWO))).toEqual(TWO)
  })

  it('空值与坏 JSON 退回空表(游客态),绝不抛', () => {
    expect(parseStoredAccounts(null)).toEqual(EMPTY_ACCOUNTS)
    expect(parseStoredAccounts('')).toEqual(EMPTY_ACCOUNTS)
    expect(parseStoredAccounts('not json')).toEqual(EMPTY_ACCOUNTS)
    expect(parseStoredAccounts('[]')).toEqual(EMPTY_ACCOUNTS)
  })

  it('坏账号剔除、重复 uid 去重、currentUid 失效时落到第一个', () => {
    const raw = JSON.stringify({
      accounts: [
        account('1001'),
        { uid: '1002' }, // 缺字段
        account('1001', { cid: 'dup' }), // 重复 uid
        account('1003'),
      ],
      currentUid: '9999',
    })
    const state = parseStoredAccounts(raw)
    expect(state.accounts.map((a) => a.uid)).toEqual(['1001', '1003'])
    expect(state.accounts[0]?.cid).toBe('cid-1001')
    expect(state.currentUid).toBe('1001')
  })
})

File: app/src/test/kotlin/com/chasel/ng2n/data/account/AccountsTest.kt
package com.chasel.ng2n.data.account

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertSame
import kotlin.test.assertTrue

class AccountsTest {

  private val two = AccountsState(
    accounts = listOf(testAccount("1001"), testAccount("1002")),
    currentUid = "1001",
  )

  @Test
  fun `新账号追加到末尾并立即成为当前账号`() {
    val state = addAccount(EMPTY_ACCOUNTS, testAccount("1001"))
    assertEquals(listOf("1001"), state.accounts.map { it.uid })
    assertEquals("1001", state.currentUid)

    val next = addAccount(state, testAccount("1002"))
    assertEquals(listOf("1001", "1002"), next.accounts.map { it.uid })
    assertEquals("1002", next.currentUid)
  }

  @Test
  fun `同 uid 重登就地刷新 cid 与 loginAt 位置不动`() {
    val relogin = testAccount("1001", cid = "cid-new", loginAt = 1_754_100_000_000)
    val state = addAccount(two, relogin)
    assertEquals(listOf("1001", "1002"), state.accounts.map { it.uid })
    assertEquals(relogin, state.accounts[0])
    assertEquals("1001", state.currentUid)
  }

  @Test
  fun `切换只动 currentUid uid 不在表里原样返回`() {
    assertEquals("1002", switchAccount(two, "1002").currentUid)
    assertSame(two, switchAccount(two, "9999"))
  }

  @Test
  fun `退出当前账号落到剩余第一个 全退光即游客态`() {
    val one = removeAccount(two, "1001")
    assertEquals(listOf("1002"), one.accounts.map { it.uid })
    assertEquals("1002", one.currentUid)

    val none = removeAccount(one, "1002")
    assertEquals(EMPTY_ACCOUNTS, none)
    assertNull(none.currentUid)
  }

  @Test
  fun `退出非当前账号不影响当前账号`() {
    assertEquals("1001", removeAccount(two, "1002").currentUid)
  }

  @Test
  fun `currentAccountOf 取当前账号 游客态是 null`() {
    assertEquals("1001", currentAccountOf(two)?.uid)
    assertNull(currentAccountOf(EMPTY_ACCOUNTS))
  }

  private val three = AccountsState(
    accounts = listOf(testAccount("1001"), testAccount("1002"), testAccount("1003")),
    currentUid = "1002",
  )

  @Test
  fun `前后循环取 到头绕回`() {
    assertEquals("1003", cycleAccountUid(three, 1))
    assertEquals("1001", cycleAccountUid(three, -1))
    assertEquals("1001", cycleAccountUid(three.copy(currentUid = "1003"), 1))
    assertEquals("1003", cycleAccountUid(three.copy(currentUid = "1001"), -1))
  }

  @Test
  fun `不足两个账号没得切`() {
    assertNull(cycleAccountUid(EMPTY_ACCOUNTS, 1))
    assertNull(
      cycleAccountUid(
        AccountsState(accounts = listOf(testAccount("1001")), currentUid = "1001"),
        1,
      ),
    )
  }

  private val day = 24L * 60 * 60 * 1000
  private val loginAt = 1_754_000_000_000

  @Test
  fun `刚登录 30 天 过一天半剩 29`() {
    assertEquals(30, cookieDaysLeft(loginAt, loginAt))
    assertEquals(29, cookieDaysLeft(loginAt, loginAt + day * 3 / 2))
    assertEquals(1, cookieDaysLeft(loginAt, loginAt + day * 59 / 2))
  }

  @Test
  fun `过了 30 天算已过期`() {
    assertTrue(cookieDaysLeft(loginAt, loginAt + 31 * day) <= 0)
    assertEquals("已过期", formatCookieExpiry(loginAt, loginAt + 31 * day))
    assertEquals("30 天后过期", formatCookieExpiry(loginAt, loginAt))
  }

  @Test
  fun `坏账号剔除 重复 uid 去重 currentUid 失效时落到第一个`() {
    val state = sanitizeAccounts(
      AccountsState(
        accounts = listOf(
          testAccount("1001"),
          testAccount("", cid = "no-uid"),
          testAccount("1002", cid = ""),
          testAccount("1001", cid = "dup"),
          testAccount("1003"),
        ),
        currentUid = "9999",
      ),
    )
    assertEquals(listOf("1001", "1003"), state.accounts.map { it.uid })
    assertEquals("cid-1001", state.accounts[0].cid)
    assertEquals("1001", state.currentUid)
  }
}
